hlink/linking/model_exploration/link_step_train_test_models.py

```diff
@@ -42,6 +151,317 @@ def __init__(self, task) -> None:
+        for index, params_combo in enumerate(all_model_parameter_combos, 1):
+            eval_start_info = f"Starting run {index} of {len(all_model_parameter_combos)} with these parameters: {params_combo}"
+            # print(eval_start_info)
+            logger.info(eval_start_info)
+            # Copy because the params combo will get stripped of extra key-values
+            # so only the hyperparams remain.
+            hyperparams = params_combo.copy()
+
+            model_type = hyperparams.pop("type")
+
+            # While we're not using thresholds in this function, we need to capture them here
+            # since they can be different for different model types and
+            # we need to use model_type, params, score and thresholds to
+            # do the next step using thresholds.
+            threshold, threshold_ratio = self._get_thresholds(
+                hyperparams, config, training_conf
+            )
+            # thresholds and model_type  are mixed in with the model hyper-parameters
+            # in the config; this removes them before passing to the model training.
+            hyperparams.pop("threshold", None)
+            hyperparams.pop("threshold_ratio", None)
+
+            pr_auc_values = self._collect_inner_kfold_cv(
+                inner_folds, model_type, hyperparams, dep_var, id_a, id_b
+            )
+            score = self._score_inner_kfold_cv_results(pr_auc_values, "mean")
+
+            model_eval = ModelEval(
+                model_type=model_type,
+                score=score,
+                hyperparams=hyperparams,
+                threshold=threshold,
+                threshold_ratio=threshold_ratio,
+            )
+            info = f"{index}: {model_eval.print()}"
+            print(info)
+            logger.debug(info)
+            results.append(model_eval)
+        return results
+
+    # Grabs the threshold settings from a single model parameter combination row (after all combinations
+    # are exploded.) Does not alter the params structure.)
+    def _get_thresholds(
+        self, model_parameters, config, training_conf
+    ) -> tuple[Any, Any]:
+        alpha_threshold = model_parameters.get(
+            "threshold", config[training_conf].get("threshold", 0.8)
+        )
+        if (
+            config[training_conf].get("decision", False)
+            == "drop_duplicate_with_threshold_ratio"
+        ):
+            threshold_ratio = model_parameters.get(
+                "threshold_ratio",
+                threshold_core.get_threshold_ratio(
+                    config[training_conf], model_parameters
+                ),
+            )
+        else:
+            threshold_ratio = False
+
+        return alpha_threshold, threshold_ratio
+
+    # Note: Returns only one model training session; if
+    # your config specified more than one model type and thresholds, you'll get
+    # the best result according to the scoring system, not the best for each
+    # model type.
```

Comment:
This makes sense to me, and should work well with how I implemented randomized parameter search.